quickwit/quickwit-proto/src/types/shard_id.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright (C) 2024 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::fmt;
+use std::fmt::Debug;
+
+use bytestring::ByteString;
+use prost::DecodeError;
+use serde::{Deserialize, Serialize};
+use ulid::Ulid;
+
+/// Shard ID.
+#[derive(Clone, Debug, Default, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub struct ShardId(ByteString);
```

Comment:
Would it be possible to keep it `Copy` with `ShardId(Ulid)` or `Shard([u8; 16])`?

Reply:
I did, but I want to eventually replace `PartitionId` with `ShardId` since they're essentially the same thing. The alternative is to use an enum:
```rust
enum ShardId {
  I64(), // Kafka
  Ulid(), // Ingest API
  String(ByteString), // Kafka, Kinesis
}
```

WDYT?

Reply:
The Protobuf serialization won't be too bad using a `oneof`.